icefoganalytics/travel-authorization
# Dockerize All Development Services/Fixup App Boot System/Fix Up Build Pipeline

Fixes https://github.com/icefoganalytics/travel-authorization/issues/27
Fixes https://github.com/icefoganalytics/travel-authorization/issues/64
Fixes https://github.com/icefoganalytics/travel-authorization/issues/22

# Context

The current app is running three services independently, the web and api services are running natively on the developer's machine. This can cause conflicts during deploy, as some code dependencies may depend on OS level local installs.
A more resilient way is to dockerize all the services, so you have a completely reproducible development environment.

See https://github.com/icefoganalytics/elcc-data-management/blob/main/docker-compose.development.yaml for an example of a dockerized development setup.

---

Currently there are two completely different systems between development and production, this is bad as the production system never gets fully tested in development.

A better system might be to have a main boot script, followed by an initialization pipeline ran via ts-node in development and node in production. Some architecting and tinkering will be needed. I'm having a hard time abstracting the calls between development and production. Development needs to run *.ts files and production needs to run *.js files.
e.g.
`ts-node-dev -r tsconfig-paths/register src/index.ts`
vs.
`node ./dist/index.js`

Make migrations run one at a time. See https://github.com/ytgov/travel-authorization/pull/53

---

Current build pipeline is very rough, and among other things, it is running the production server in development mode.
See https://github.com/icefoganalytics/elcc-data-management/blob/main/Dockerfile for an example of a multi-stage build that handles typescript build requirements, and runs in production mode.

The README is also out of date regarding the build system.

# Implementation

Rebuild production Dockerfile as multi-stage build that compiles in development mode for Typescript in initial stages, then installs packages and runs in production mode in final stage.
Dockerise web service.
Merge development and production app boot scripts.
Make migrations run one at a time.
Move migration running to an initializer pipeline.
Use environment specific database names for local builds to avoid conflicts around migration names.
Move a bunch of environment variables into the docker-compose.development.yaml.

# Testing Instructions

1. Boot the app cleanly in development mode via `dev down -v && dev up --build`.
2. Check that the web, api, and db service all come up.
3. Check that you can log in to the app at http://localhost:8080/
4. Check that you can go to the "My Travel Request" page from the top nav.
5. Check that you can create a travel request via the "+ Travel Authorization" button.
---
1. Check that you can boot the app in production mode by following the instructions in the README.
    > mostly this involves setting up the right values in the .env file.
2. Build the production setup via `docker compose down -v && docker compose up --build`. It will take a while.
3. Check that you can log in to the app at http://localhost:3000/
4. Check that you can go to the "My Travel Request" page from the top nav.
5. Check that you can create a travel request via the "+ Travel Authorization" button.
